### DISCUSSION TITLE
[DISCUSSION] Appwrite Messaging API - Community feedback is required*

### DISCUSSION THREAD

**Discussion by miguelramos** [appwrite/appwrite#905] [Ideas]
@eldadfux Not only open issues but to give you 3 ideas for new features. Push notifications, email sender and messaging. Perhaps they are huge but could give lot of visibility to the project.

Regards, :).

**Comment by eldadfux** [comment]
Hey @miguelramos, this a great idea that was in Appwrite original blueprint.

We left it out of our first versions because implementing it had a few issues we didn't have time to address.

It's something we definitely would like to come back to in the future!

**Comment by miguelramos** [comment]
Cool, maybe mosquito(MQTT) for the infra and it is well know in IOT world and mailhog/maildev too, but is just a consideration and yes will increase the infrastructure but they are cool options.

**Comment by eldadfux** [comment]
Yeh, no doubt, this is a massive feature in terms of infrastructure and user integration. I would love to get some more community feedback on this one.

Our initial idea was to create a messaging API that has multiple adapters for sending messages to different endpoints.

Some of the adapters we had in mind included:

* **Email** - Using Appwrite current SMTP server or a user-defined one (will have to find a good solution for SPAM preventing)
* **Push Messages** - Should probably support at lease iOS, Android and Web
* **Telegram** - Probably the easiest one to implement
* **SMS** - will require external provider, might also allow us to use this for future 2fa feature

Some of the main concerns I have, including but not limited to are:
* How to easily allow developers to integrate correctly with adapters
* The infrastructure overhead - we need to keep default setup cheap (5-10$ servers)
* Making sure the API is consistent and predictable
* How should the dashboard look like, should we have stats alongside adapters configuration, what else is needed in the console?

**Comment by miguelramos** [comment]
As initial point i agree that this should be an API. It will facilitate first and third point in concerns list. For push messages and to not increase costs and maintain the actual budget you could test: https://gotify.net/ seems a pretty good thing and start from there. For email will be a huge challenge, maybe someone in the community have experience and know-how for doing this and bring what we need to prevent spam and other traps.

**Comment by wentzlau** [comment]
There are two kinds of messaging that Appwrite should support. 

1. User notification via push messages on different communication channels.
2. Internal application bus. A hi level bus api where the underlying broker is hidden and where is it possible to post events, queries and commands (CQRS)

As we need to subscripe for push messages in our applications it requires a websocket connection point in Appwrite.

**Reply by stnguyen90** [reply]
Would you be able to expand more on the 2nd point? What would it be used for?

**Comment by eldadfux** [comment]
@wentzlau great feedback, the push messaging is definitely something we'll have on our roadmap.

I also really  like the idea of having dedicated support for server-to-server communication. Do you have any suggestions for a communication protocol that will be good for this type of communication?

**Comment by chansdad** [comment]
> As initial point i agree that this should be an API. It will facilitate first and third point in concerns list. For push messages and to not increase costs and maintain the actual budget you could test: https://gotify.net/ seems a pretty good thing and start from there. For email will be a huge challenge, maybe someone in the community have experience and know-how for doing this and bring what we need to prevent spam and other traps.

Gotify i think only supports plain text and has no support for Rich Text messages. OneSignal was great , but it is not longer free and opensource .

**Comment by Saad-ISAA** [comment]
THough it's a very early discussion here, but i was looking to push notifications just like firebase has if it was possible to do push notifications to mobile devices from appwrite.
There should be an API for it with the sdk so that it can be used programatically to send push notifications.
Overall Appwrite has almost all the features of Google's Firebase, but this main feature is still remaining. I hope this feature is included soon

**Comment by simpros** [comment]
Are there any news on this? This missing feature is the only thing why I would prefer parse-platform over appwrite.

**Comment by YouSour** [comment]
wake me up if you guys have any news about this 😄

**Comment by squallsama** [comment]
Push notifications with scheduled rules would be awesome feature. This is a last thing that keeps me using firebase instead of appwrite.

**Comment by p4-k4** [comment]
Placing my vote.

**Comment by glin94** [comment]
Me too

**Comment by dasaintgray** [comment]
That's the feature that I've been waiting for, I hope that Appwrite team will do that.  I'm placing my vote to include push notifications.

**Comment by stevebrainng** [comment]
Good day to the Appwrite team, it looks like this feature is now abandoned or still no plans to have any of this feature anymore. This feature is one of Appwrite most demanded feature. any plans and updates?

**Comment by nbehler13** [comment]
I would realy love seeing a push notification feature! @stnguyen90 can you give some insights if this feature will be added? :)

**Reply by squallsama** [reply]
You can do it via appwrite cloud functions already using FCM

**Reply by p4-k4** [reply]
> You can do it via appwrite cloud functions already using FCM

Since when was Firebase Cloud Messaging an appwrite feature?

**Reply by mustansirgodhrawala** [reply]
FCM is not a firebase feature, it is a android feature. Push notifications can't be done via the programmer, they're handled by the operating system creators, FCM for android and APNS for apple. At the end of the day even appwrite will have to integrate with FCM and APN to support push notifications. Essentially requiring the developer to register with firebase and upload a service account key to appwrite anyways, and upload a apple p8 key for supporting IOS.

Edit: Appwrite functions, to support push notifications makes way more sense than at first thought, albeit a little inconvenient. Gives you way more control, allows the open source devs to not have to recreate a lot of the underlying code used by the functions feature anyways.

**Reply by sergioponguta** [reply]
> FCM is not a firebase feature, it is a android feature. Push notifications can't be done via the programmer, they're handled by the operating system creators, FCM for android and APNS for apple. At the end of the day even appwrite will have to integrate with FCM and APN to support push notifications. Essentially requiring the developer to register with firebase and upload a service account key to appwrite anyways, and upload a apple p8 key for supporting IOS.
> 
> Edit: Appwrite functions, to support push notifications makes way more sense than at first thought, albeit a little inconvenient. Gives you way more control, allows the open source devs to not have to recreate a lot of the underlying code used by the functions feature anyways.

Thanks for the useful information!

**Comment by devmarwat** [comment]
@eldadfux Push Notifications 🙏

**Comment by p4-k4** [comment]
Well, this idea seems a tad abandoned.

**Reply by Kilo-Loco** [reply]
It seems like there may be some updates coming soon: https://github.com/appwrite/appwrite/discussions/1623#discussioncomment-8227703
